pkg/controller/sandbox/sandbox_controller.go

```diff
@@ -139,58 +115,49 @@ func (r *SandboxReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return reconcile.Result{}, err
 	} else if errors.IsNotFound(err) {
 		pod = nil
-	} else if !pod.DeletionTimestamp.IsZero() {
-		return reconcile.Result{RequeueAfter: min(requeueAfter, time.Second*3)}, nil
-	} else if pod.Status.Phase == corev1.PodSucceeded {
-		newStatus.Phase = agentsv1alpha1.SandboxSucceeded
-		return ctrl.Result{RequeueAfter: requeueAfter}, r.updateSandboxStatus(ctx, *newStatus, box)
-	} else if pod.Status.Phase == corev1.PodFailed &&
-		box.Status.Phase != agentsv1alpha1.SandboxPaused {
-		// During the paused phase, the pod transitions to the Failed state and should be ignored.
-		newStatus.Phase = agentsv1alpha1.SandboxFailed
-		return ctrl.Result{RequeueAfter: requeueAfter}, r.updateSandboxStatus(ctx, *newStatus, box)
 	}
+	args := core.EnsureFuncArgs{Pod: pod, Box: box, NewStatus: newStatus}
 
+	// ensure sandbox terminating
 	if !box.DeletionTimestamp.IsZero() {
-		newStatus.Phase = agentsv1alpha1.SandboxTerminating
-		cond := utils.GetSandboxCondition(newStatus, string(agentsv1alpha1.SandboxConditionReady))
-		if cond != nil && cond.Status == metav1.ConditionTrue {
-			cond.Status = metav1.ConditionFalse
-			cond.LastTransitionTime = metav1.Now()
-			utils.SetSandboxCondition(newStatus, *cond)
-		}
-		// If it is paused, first set the sandbox to the Paused state.
-		// To prevent loss of state information, the state immediately before Paused must currently be Running.
-	} else if box.Spec.Paused && box.Status.Phase == agentsv1alpha1.SandboxRunning {
-		// The paused and resumed condition are exclusive
-		utils.RemoveSandboxCondition(newStatus, string(agentsv1alpha1.SandboxConditionResumed))
-		newStatus.Phase = agentsv1alpha1.SandboxPaused
-		// enter resume phase
-	} else if !box.Spec.Paused && newStatus.Phase == agentsv1alpha1.SandboxPaused {
-		// delete paused condition
-		utils.RemoveSandboxCondition(newStatus, string(agentsv1alpha1.SandboxConditionPaused))
-		newStatus.Phase = agentsv1alpha1.SandboxResuming
-		cond := metav1.Condition{
-			Type:               string(agentsv1alpha1.SandboxConditionResumed),
-			Status:             metav1.ConditionFalse,
-			Reason:             agentsv1alpha1.SandboxResumeReasonCreatePod,
-			LastTransitionTime: metav1.Now(),
-		}
-		utils.SetSandboxCondition(newStatus, cond)
+		return r.handleTerminating(ctx, args)
+	}
+
+	// if sandbox phase = Failed, Success
+	if r.isTerminalPhase(box.Status.Phase) {
+		return ctrl.Result{}, nil
+	}
+
+	// add finalizer
+	if box, err = r.ensureFinalizer(ctx, box); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check Shutdown
+	requeueAfter, err := r.handleShutdownTime(ctx, box)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
```

Comment:
consider check requeueAfter and return, calculateStatus may return shouldRequeue=false, so that the requeue will not happen at all if shutdown is near